### PR TITLE
Make 0 an usable count for commands

### PR DESCRIPTION
**Pull request by knaggita**

### CODE REVIEW

**Comment by The-Compiler** [comment]
Now the function handler has a `count` and a `zero_count` argument, which is confusing :wink:

It'd be better to have something like `@cmdutils.argument('count', count=True, allow_zero=True)` or so, so the normal `count` can be used like before.

File: qutebrowser/browser/commands.py [hunk 1]
```diff
@@ -968,6 +968,8 @@ def tab_focus(self, index: typing.Union[str, int]=None, count=None):
                    Negative indices count from the end, such that -1 is the
                    last tab.
             count: The tab index to focus, starting with 1.
+            zero_count: The tab index to focus, with 0. The special value
+                   0 value focuses the last focused tab.
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: This should be in the description for `count` now, `zero_count` doesn't exist as argument to `tab_focus` anymore.

File: qutebrowser/browser/commands.py [hunk 2]
```diff
@@ -976,9 +978,11 @@ def tab_focus(self, index: typing.Union[str, int]=None, count=None):
         if index is None:
             self.tab_next()
             return
+        if index == 0:
+            self._tab_focus_last()
+            return
         if index < 0:
             index = self._count() + index + 1
-
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: Please add this back :wink:

File: qutebrowser/commands/command.py [hunk 3]
```diff
@@ -33,12 +33,13 @@ class ArgInfo:
 
     """Information about an argument."""
 
-    def __init__(self, win_id=False, count=False, flag=None, hide=False,
-                 metavar=None, completion=None, choices=None):
+    def __init__(self, win_id=False, count=False, hide=False, metavar=None,
+                 zero_count=False, flag=None, completion=None, choices=None):
         if win_id and count:
             raise TypeError("Argument marked as both count/win_id!")
```
Comment (The-Compiler) [inline qutebrowser/commands/command.py]: Similar to this, we should fail if `zero_count=True` is given without `count=True` because that combination makes no sense.

Please also add a test for that scenario to `tests/unit/commands/test_cmdutils.py` (similar to `TestArgument.test_count_and_win_id_same_arg`).

Comment (The-Compiler) [inline qutebrowser/commands/command.py]: This is still open - what I meant is that you could accidentally do `@cmdutils.argument('count', zero_count=True)` and forget `count=True` - since `zero_count` makes no sense without `count`, that case should raise an exception and be tested.

File: tests/unit/commands/test_cmdutils.py [hunk 4]
```diff
@@ -397,6 +397,16 @@ def fun(arg=0):
...
+    def test_zero_count_and_win_id_same_arg(self):
+        with pytest.raises(TypeError) as excinfo:
+            @cmdutils.argument('arg', count=True,
+             zero_count=True, win_id=True)
+            def fun(arg=0):
+                """Blah."""
+                pass
+
+        assert str(excinfo.value) == "Argument marked as both count/win_id!"
+
```
Comment (The-Compiler) [inline tests/unit/commands/test_cmdutils.py]: See my comment above - the test should check `zero_count=True` without `count=True` - this test as-is doesn't really test anything new.

File: tests/end2end/features/tabs.feature [hunk 5]
```diff
@@ -255,6 +255,18 @@ Feature: Tab management
             - data/numbers/2.txt (active)
             - data/numbers/3.txt
 
+    Scenario: :tab-focus with 0
+        When I open data/numbers/1.txt
+        And I open data/numbers/2.txt in a new tab
+        And I open data/numbers/3.txt in a new tab
+        And I run :tab-focus 1
+        And I run :tab-focus 3
+        And I run :tab-focus 0
```
Comment (The-Compiler) [inline tests/end2end/features/tabs.feature]: This tests the index *argument* to `:tab-focus`, which is different from the count (which gets passed from keybindings, e.g. `0T`).

You can use `And I run :tab-focus with count 0` to test the count.

File: tests/end2end/features/tabs.feature [hunk 6]
```diff
@@ -255,6 +255,18 @@ Feature: Tab management
...
+        Then the following tabs should be open:
+            - data/numbers/1.txt (active)
```
Comment (The-Compiler) [inline tests/end2end/features/tabs.feature]: I find this surprising - wouldn't we expect `:tab-focus` with count (or index) 0 to focus the *last* tab now?

File: tests/unit/keyinput/test_basekeyparser.py [hunk 7]
```diff
@@ -314,9 +315,8 @@ def test_no_count(self, handle_text, keyparser):
 
     def test_count_0(self, handle_text, keyparser):
         handle_text((Qt.Key_0, '0'), (Qt.Key_B, 'b'), (Qt.Key_A, 'a'))
-        calls = [mock.call('0', keyparser.Type.chain, None),
-                 mock.call('ba', keyparser.Type.chain, None)]
-        keyparser.execute.assert_has_calls(calls)
+        keyparser.execute.assert_called_once_with(
+            'ba', keyparser.Type.chain, 0)
```
Comment (The-Compiler) [inline tests/unit/keyinput/test_basekeyparser.py]: Good job on adjusting the tests here, I totally forgot about them! :+1:

File: qutebrowser/browser/commands.py [hunk 8]
```diff
@@ -955,7 +955,7 @@ def buffer(self, index):
 
     @cmdutils.register(instance='command-dispatcher', scope='window')
     @cmdutils.argument('index', choices=['last'])
-    @cmdutils.argument('count', count=True)
+    @cmdutils.argument('count', count=True, zero_count=True)
     def tab_focus(self, index: typing.Union[str, int]=None, count=None):
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: Just to reiterate on this again, because I think this is confusing you: `:tab-focus` can accept a tab number either as argument `index` (like `:tab-focus 1`) or as a count via a keybinding (e.g. when pressing `0T`, then `index` would be `None` but `count` would be `0`).

Then the `index = count if count is not None else index` line below overrides the index with the count if it's given.

File: qutebrowser/browser/commands.py [hunk 9]
```diff
@@ -968,6 +968,8 @@ def tab_focus(self, index: typing.Union[str, int]=None, count=None):
                    Negative indices count from the end, such that -1 is the
                    last tab.
             count: The tab index to focus, starting with 1.
+                   zero_count: The tab index to focus, with 0.
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: This line doesn't make any sense anymore here.

File: qutebrowser/browser/commands.py [hunk 10]
```diff
@@ -976,6 +978,9 @@ def tab_focus(self, index: typing.Union[str, int]=None, count=None):
         if index is None:
             self.tab_next()
             return
+        if index == 0:
+            self._tab_focus_last()
+            return
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: The idea was to focus the rightmost tab with `0T`, not the last focused one - the docstring above should then also be adjusted accordingly.

File: qutebrowser/commands/command.py [hunk 11]
```diff
@@ -173,6 +176,9 @@ def _check_prerequisites(self, win_id):
                 "{}: Only available with {} "
                 "backend.".format(self.name, self.backend.name))
 
+        if count == 0:
+            raise cmdexc.PrerequisitesError("Zero count not allowed")
```
Comment (The-Compiler) [inline qutebrowser/commands/command.py]: Please add the command name at the beginning of the error message here, like with the other errors.

File: qutebrowser/commands/command.py [hunk 12]
```diff
@@ -173,6 +176,9 @@ def _check_prerequisites(self, win_id):
                 "{}: Only available with {} "
                 "backend.".format(self.name, self.backend.name))
 
+        if count == 0:
```
Comment (The-Compiler) [inline qutebrowser/commands/command.py]: The missing bit here is that we actually want to allow a 0 count when the argument has `zero_count=True` set, but we don't actually have the necessary information here.

I'd propose something like this (untested):

- In `Command.__init__`, set `self._zero_count_allowed = None` (`None` because we don't know yet), and describe the argument in the docstring
- Here, do `if count == 0 and not self._zero_count_allowed:`
- Now we need to find out which argument is the count, and then set `self._zero_count_allowed` according to that - the best place to do that is probably in `_inspect_func`, in the for loop running for every parameter. There you can get the `ArgInfo` object using `self._get_arg_info(param)`, and if `arg_info.count` is true (i.e. we're looking at the `count` argument), set `self._zero_count_allowed = arg_info.zero_count`

I hope that made sense? If not, feel free to ask! :smile:

File: qutebrowser/browser/commands.py [hunk 13]
```diff
@@ -976,6 +977,9 @@ def tab_focus(self, index: typing.Union[str, int]=None, count=None):
         if index is None:
             self.tab_next()
             return
+        if index == 0:
+            index = self._count()
+
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: Not something directly related to your change, but I think it'd be cleaner to combine this into an `if`/`elif`/`elif` with the check above/below, as they're mutually exclusive - could you please do that?

File: qutebrowser/commands/command.py [hunk 14]
```diff
@@ -33,12 +33,15 @@ class ArgInfo:
...
     """Information about an argument."""
 
-    def __init__(self, win_id=False, count=False, flag=None, hide=False,
-                 metavar=None, completion=None, choices=None):
+    def __init__(self, win_id=False, count=False, hide=False, metavar=None,
+                 zero_count=False, flag=None, completion=None, choices=None):
         if win_id and count:
             raise TypeError("Argument marked as both count/win_id!")
+        if zero_count and not count:
+            raise TypeError("Zero_count Argument cannot exist without count!")
```
Comment (The-Compiler) [inline qutebrowser/commands/command.py]: nitpick: Please lower-case `Zero_count` (as it's a literal argument name) and `Argument` here.

File: qutebrowser/commands/command.py [hunk 15]
```diff
@@ -173,6 +179,11 @@ def _check_prerequisites(self, win_id):
                 "{}: Only available with {} "
                 "backend.".format(self.name, self.backend.name))
 
+        if count == 0 and not self._zero_count:
+            raise cmdexc.PrerequisitesError(
+                "{}: Argument zero_count not been set to true!".format(
+                    self.name))
```
Comment (The-Compiler) [inline qutebrowser/commands/command.py]: This is the error displayed to the user, so this would be a confusing output. I think something like "{}: A zero count is not allowed for this command!" would be better.

File: tests/end2end/features/tabs.feature [hunk 16]
```diff
@@ -255,6 +255,16 @@ Feature: Tab management
             - data/numbers/2.txt (active)
             - data/numbers/3.txt
 
+    Scenario: :tab-focus with 0
+        When I open data/numbers/1.txt
+        And I open data/numbers/2.txt in a new tab
+        And I open data/numbers/3.txt in a new tab
+        And I run :tab-focus 0
```
Comment (The-Compiler) [inline tests/end2end/features/tabs.feature]: This still checks the argument and not the count, it should probably be "And I run :tab-focus with count 0" instead.

Also, the last tab is already focused by default, so I guess you'd need to do "And I run :tab-focus 1" or so first for the test to be effective.

Comment (The-Compiler) [inline tests/end2end/features/tabs.feature]: This still checks the argument and not the count, it should probably be "And I run :tab-focus with count 0" instead.

Also, the last tab is already focused by default, so I guess you'd need to do "And I run :tab-focus 1" or so first for the test to be effective.

File: tests/unit/commands/test_cmdutils.py [hunk 17]
```diff
@@ -397,6 +397,17 @@ def fun(arg=0):
 
         assert str(excinfo.value) == "Argument marked as both count/win_id!"
 
+    def test_zero_count_and_zero_count_arg(self):
+        with pytest.raises(TypeError) as excinfo:
+            @cmdutils.argument('arg', count=False,
+             zero_count=True)
```
Comment (The-Compiler) [inline tests/unit/commands/test_cmdutils.py]: This should fit on the line above without going over the 79-char limit, no?

File: tests/unit/commands/test_cmdutils.py [hunk 18]
```diff
@@ -397,6 +397,17 @@ def fun(arg=0):
...
+            def fun(arg=0):
+                """Blah."""
+                pass
+
+        assert str(excinfo.value) == ("Zero_count Argument" +
```
Comment (The-Compiler) [inline tests/unit/commands/test_cmdutils.py]: I'd prefer splitting this up via a variable:

```python
        expected = "zero_count argument cannot exist without count!"
        assert str(excinfo.value) == expected
```

**Comment by The-Compiler** [comment]
Small update: I wanted to look at this this morning, but my internet connection in the train was horrible - I have it pulled locally now though, so I hope I'll get to merging this later today.

**Comment by The-Compiler** [comment]
Squashed and merged as e73932a555bbfd6e391dbd65977253bba592b83c with a few changes on top:

- 6b76d5defa639e6222a8c2e902a61834f25c79b2 with some whitespace adjustments
- 083d84731627b81bbcab0e2b7979c0399359ea90 adding a test for a command which doesn't support a zero count
- c80f18522c11396a1551569b6cae11eae844ba22 removing a now duplicated test

Thanks!
